### PR TITLE
Indexing data

### PR DESCRIPTION
**Pull request by cableman**
#### Link to ticket

https://jira.itkdev.dk/browse/EV-268

#### Description

Index event and daily occurrences into a search index.

#### Screenshot of the result

N/A

#### Checklist

- [ ] My code is covered by test cases.
- [x] My code passes our test (all our tests).
- [x] My code passes our static analysis suite.
- [x] My code passes our continuous integration process.

If your code does not pass all the requirements on the checklist you have to add a comment explaining why this change 
should be exempt from the list.

#### Additional comments or questions

N/A

### CODE REVIEW

File: symfony.lock [hunk 1]
```diff
@@ -84,6 +84,18 @@
             "config/packages/nelmio_cors.yaml"
         ]
     },
+    "php-http/discovery": {
```
Comment (cableman) [inline symfony.lock:87]: This extension comes from running an `composer update`

File: src/MessageHandler/GeocoderHandler.php [hunk 2]
```diff
@@ -34,7 +33,7 @@ public function __invoke(GeocoderMessage $message): void
             try {
                 $coordinates = $this->geocoderService->encode($address);
                 $address->setLatitude($coordinates[0]);
-                $address->setLongitude($coordinates[0]);
+                $address->setLongitude($coordinates[1]);
```
Comment (rimi-itk) [inline src/MessageHandler/GeocoderHandler.php:36]: 🎉

File: src/Model/Indexing/IndexFieldTypes.php [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace App\Model\Indexing;
+
+final class IndexFieldTypes
+{
+    // ISO 8601 date "2004-02-12T15:19:21+00:00"
+    public const DATEFORMAT = 'c';
```
Comment (rimi-itk) [inline src/Model/Indexing/IndexFieldTypes.php:8]: Consider using [ISO8601_EXPANDED ](https://www.php.net/manual/en/class.datetimeinterface.php#datetimeinterface.constants.iso8601-expanded) or [ATOM](https://www.php.net/manual/en/class.datetimeinterface.php#datetimeinterface.constants.atom) (both are easier to read and understand the meaning of than `c`):
```suggestion
    public const DATEFORMAT = \DateTimeInterface::ISO8601_EXPANDED;
```

File: src/Model/Indexing/IndexFieldTypes.php [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+    // One would think one could use the named ISO formats (but, NO) and Elastic say they use the format in the link.
+    // Here it also needs to be "z" and not "Z" to match the ISO format.
+    // @see https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html
+    public const DATEFORMAT_ES = "yyyy-MM-dd'T'HH:mm:ssz";
```
Comment (rimi-itk) [inline src/Model/Indexing/IndexFieldTypes.php:13]:
```suggestion
    public const DATEFORMAT_ES = 'yyyy-MM-dd'T'HH:mm:ssz';
```

File: src/Model/Indexing/IndexFieldTypes.php [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace App\Model\Indexing;
+
+final class IndexFieldTypes
+{
+    // ISO 8601 date "2004-02-12T15:19:21+00:00"
+    public const DATEFORMAT = 'c';
+
+    // One would think one could use the named ISO formats (but, NO) and Elastic say they use the format in the link.
```
Comment (rimi-itk) [inline src/Model/Indexing/IndexFieldTypes.php:10]: Which format do Elastic say they actually use? https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_INSTANT? The provided link points to broad information on how to format dates in Java.

File: src/Model/Indexing/IndexFieldTypes.php [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    // Here it also needs to be "z" and not "Z" to match the ISO format.
```
Comment (rimi-itk) [inline src/Model/Indexing/IndexFieldTypes.php:11]:
```suggestion
    // Here it also needs to be "z" (“time-zone name”) and not "Z" (“zone-offset”) to match the ISO format.
```

File: src/Model/Indexing/IndexNames.php [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+namespace App\Model\Indexing;
+
+enum IndexNames: string
+{
+    case Events = 'events';
+    case Organization = 'organization';
+
+    public static function values(): array
+    {
+        return array_column(IndexNames::cases(), 'value');
```
Comment (rimi-itk) [inline src/Model/Indexing/IndexNames.php:12]: This is clever! And I didn't know (until now) that you can use `array_column` on objects. Cool!

File: src/Service/ImageHandlerInterface.php [hunk 8]
```diff
@@ -6,9 +6,45 @@
...
+     * @return string
+     *   Relative path to the downloaded file
+     */
     public function fetch(string $url): string;
 
+    /**
+     * Remove image from the filesystem.
+     *
+     * @param Image $image
+     *   The image top remove
```
Comment (rimi-itk) [inline src/Service/ImageHandlerInterface.php:24]:
```suggestion
     *   The image to remove
```

File: src/Service/ImageHandlerInterface.php [hunk 9]
```diff
@@ -6,9 +6,45 @@
...
+     *
+     * @return bool
+     *   True if image was removed else false
+     */
     public function remove(Image $image): bool;
 
+    /**
+     * Make image transformations.
+     *
+     * @param image $image
```
Comment (rimi-itk) [inline src/Service/ImageHandlerInterface.php:34]:
```suggestion
     * @param Image $image
```

File: src/Service/ImageHandlerInterface.php [hunk 10]
```diff
@@ -6,9 +6,45 @@
...
+     *   The image to make derived/transformation from
```
Comment (rimi-itk) [inline src/Service/ImageHandlerInterface.php:35]:
```suggestion
     *   The image to transform
```

File: src/Service/ImageHandlerInterface.php [hunk 11]
```diff
@@ -6,9 +6,45 @@
...
+     */
     public function transform(Image $image): void;
+
+    /**
+     * Get urls for derived images from image url.
```
Comment (rimi-itk) [inline src/Service/ImageHandlerInterface.php:40]: Is a “derived image” a “transformed image”? If so, use “transformed” to make it more clear.

File: src/Service/ImageHandlerInterface.php [hunk 12]
```diff
@@ -6,9 +6,45 @@
...
+     *
+     * @param string $imageUrl
+     *   The local url for the image
```
Comment (rimi-itk) [inline src/Service/ImageHandlerInterface.php:43]: Is “The local url for the image” the one returned by `fetch`? If so, rename to `$imagePath` and make a note on this.

File: src/Service/ImageHandlerInterface.php [hunk 13]
```diff
@@ -6,9 +6,45 @@
...
+     * @param bool $absolute
+     *   If true absolute path
+     *
+     * @return array<string>
+     */
+    public function getDerived(string $imageUrl, bool $absolute = true): array;
```
Comment (rimi-itk) [inline src/Service/ImageHandlerInterface.php:49]:
```suggestion
    public function getTransformedImagePaths(string $imagePath, bool $absolute = true): array;
```

File: src/Service/Indexing/AbstractIndexingElastic.php [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+            'index' => $this->indexAliasName,
+            'id' => $item->getId(),
+            'body' => $this->serialize($item),
+        ];
+
+        try {
+            /** @var Elasticsearch $response */
+            $response = $this->client->index($params);
+
+            if (Response::HTTP_OK !== $response->getStatusCode() && Response::HTTP_CREATED !== $response->getStatusCode() && Response::HTTP_NO_CONTENT !== $response->getStatusCode()) {
```
Comment (rimi-itk) [inline src/Service/Indexing/AbstractIndexingElastic.php:35]:
```suggestion
            if (!in_array($response->getStatusCode(), [Response::HTTP_OK, Response::HTTP_CREATED, Response::HTTP_NO_CONTENT])) {
```

File: src/Service/Indexing/AbstractIndexingElastic.php [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+        $params = [
+            'index' => $this->indexAliasName,
+            'id' => $id,
+        ];
+
+        try {
+            /** @var Elasticsearch $response */
+            $response = $this->client->delete($params);
+
+            if (Response::HTTP_OK !== $response->getStatusCode() && Response::HTTP_ACCEPTED !== $response->getStatusCode() && Response::HTTP_NO_CONTENT !== $response->getStatusCode()) {
```
Comment (rimi-itk) [inline src/Service/Indexing/AbstractIndexingElastic.php:54]:
```suggestion
            if (!in_array($response->getStatusCode(), [Response::HTTP_OK, Response::HTTP_ACCEPTED, Response::HTTP_NO_CONTENT])) {
```

File: src/Service/Indexing/AbstractIndexingElastic.php [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+                    'index' => [
+                        '_index' => $this->newIndexName,
+                        '_id' => $item->getId(),
+                    ],
+                ];
+                $params['body'][] = $this->serialize($item);
+            }
+
+            $response = $this->client->bulk($params);
+            if (Response::HTTP_OK !== $response->getStatusCode() && Response::HTTP_CREATED !== $response->getStatusCode() && Response::HTTP_NO_CONTENT !== $response->getStatusCode()) {
```
Comment (rimi-itk) [inline src/Service/Indexing/AbstractIndexingElastic.php:83]:
```suggestion
            if (!in_array($response->getStatusCode(), [Response::HTTP_OK, Response::HTTP_CREATED, Response::HTTP_NO_CONTENT])) {
```

File: src/Service/Indexing/IndexingEvents.php [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+        $contextBuilder = (new ObjectNormalizerContextBuilder())
+            ->withGroups([IndexNames::Events->value]);
+        $contextBuilder = (new DateTimeNormalizerContextBuilder())
+            ->withContext($contextBuilder)
+            ->withFormat(IndexFieldTypes::DATEFORMAT);
+        $data = $this->serializer->normalize($item, null, $contextBuilder->toArray());
+
+        // @todo: Figure out how to do these changes with the serializer. This is just....
+        // Flatten tags.
+        $data['tags'] = array_map(fn ($tag) => $tag['name'], $data['tags']);
```
Comment (rimi-itk) [inline src/Service/Indexing/IndexingEvents.php:43]:
```suggestion
        // Get tag names.
        $data['tags'] = array_column($data['tags'], 'name');
```

File: src/Service/Indexing/IndexingEvents.php [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+
+        // Flatten location address and convert lang/long to coordinate point.
+        $data['location'] += $data['location']['address'];
+        unset($data['location']['address']);
+        $data['location']['coordinates'] = [];
+        $data['location']['coordinates'] = [$data['location']['latitude'], $data['location']['longitude']];
```
Comment (rimi-itk) [inline src/Service/Indexing/IndexingEvents.php:49]:
```suggestion
        $data['location']['coordinates'] = [$data['location']['latitude'], $data['location']['longitude']];
```

File: src/Service/Indexing/IndexingEvents.php [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+        unset($data['location']['latitude']);
+        unset($data['location']['longitude']);
```
Comment (rimi-itk) [inline src/Service/Indexing/IndexingEvents.php:51]:
```suggestion
        unset($data['location']['latitude'], $data['location']['longitude']);
```

File: src/Service/Indexing/IndexingEvents.php [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+
+        // Fix image urls (with a full path and derived sizes).
+        $imageUrl = $data['imageUrl']['original'];
+        $data['imageUrl'] = $this->imageHandler->getDerived($imageUrl);
```
Comment (rimi-itk) [inline src/Service/Indexing/IndexingEvents.php:55]: `src/Service/ImageHandlerInterface::getDerived` returns an array:

```suggestion
        $data['imageUrls'] = $this->imageHandler->getDerived($imageUrl);
```

File: src/Service/Indexing/IndexingOrganization.php [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+                                'format' => IndexFieldTypes::DATEFORMAT_ES,
+                                'index' => false,
+                                'doc_values' => true,
+                            ],
+                        ],
+                    ],
+                ],
+            ]);
+
+            if (Response::HTTP_OK !== $response->getStatusCode() && Response::HTTP_NO_CONTENT !== $response->getStatusCode()) {
```
Comment (rimi-itk) [inline src/Service/Indexing/IndexingOrganization.php:100]:
```suggestion
            if (!in_array($response->getStatusCode(), [Response::HTTP_OK, Response::HTTP_NO_CONTENT])) {
```

File: src/Service/Indexing/IndexingOrganization.php [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+                            ],
+                            'updated' => [
+                                'type' => 'date',
+                                'format' => IndexFieldTypes::DATEFORMAT_ES,
+                                'index' => false,
+                                'doc_values' => true,
+                            ],
+                        ],
+                    ],
+                ],
```
Comment (rimi-itk) [inline src/Service/Indexing/IndexingOrganization.php:97]: Should/can we move this to config in `services.yaml`, say?

File: src/Model/Indexing/IndexFieldTypes.php
Comment (rimi-itk) [inline src/Model/Indexing/IndexFieldTypes.php:1]: It's not obvious (to me at least) how `IndexFieldTypes::DATEFORMAT` and `IndexFieldTypes:DATEFORMAT_ES` are used.

Do we use `DATEFORMAT` to format a datetime (in PHP) and then tell Elastic that the formatted datetime (i.e. a string) is in the `DATEFORMAT_ES` format?

Comment (cableman) [reply src/Model/Indexing/IndexFieldTypes.php:1]: The formats are the same, but PHP and ES (which is Java) uses different representation for the formats.

File: src/Model/Indexing/IndexFieldTypes.php (hunk 5 above)
Comment (cableman) [reply src/Model/Indexing/IndexFieldTypes.php:10]: The default date format in ES do not include time and there predefined name for date-time contains 3 "s" for seconds. Tried to use it but it did not work with the current version of ES... found a lot of folks online complaining about the build in data formats. So ended up using the date time without the extra "s" in the format and ES accepted that format.

File: config/services.yaml [hunk 23]
```diff
@@ -22,7 +22,44 @@ services:
...
+
+    App\Service\Populate:
+        arguments:
+            $indexingServices: !tagged { tag: 'indexing_interface', index_by: 'key' }
+            $repositories: !tagged { tag: 'populate_interface', index_by: 'key' }
+
+    App\MessageHandler\IndexHandler:
+        arguments:
+            $indexingServices: !tagged { tag: 'indexing_interface', index_by: 'key' }
+            $repositories: !tagged { tag: 'populate_interface', index_by: 'key' }
```
[the remark this answers is not in the document]
Comment (cableman) [reply config/services.yaml:44]: I think this is the most explicit way to define this and the easiest to read later on.

File: src/Command/Index/IndexCreateCommand.php [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+            try {
+                if (!$service->indexExists()) {
+                    $service->createIndex();
+
+                    $io->success('Index created: '.$index);
+                } else {
+                    $io->caution('Index exists ('.$index.'). Aborting.');
+                }
+            } catch (IndexingException $e) {
+                $io->error($e->getMessage());
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Command/Index/IndexCreateCommand.php:62]: Not sure sending this to an logged would be useful. This command should only be executed once at site setup or if something has gone horribly wrong at the complete index have been lost.

File: src/Model/Indexing/IndexFieldTypes.php (hunk 3 above)
Comment (cableman) [reply src/Model/Indexing/IndexFieldTypes.php]: Fixed

File: src/Model/Indexing/IndexFieldTypes.php (hunk 4 above)
Comment (cableman) [reply src/Model/Indexing/IndexFieldTypes.php:16]: Fixed

File: src/Service/Indexing/IndexingEvents.php (hunk 20 above)
Comment (cableman) [reply src/Service/Indexing/IndexingEvents.php]: Fixed

File: src/Service/ImageHandlerInterface.php (hunk 11 above)
Comment (cableman) [reply src/Service/ImageHandlerInterface.php:40]: Fixed

File: src/Service/ImageHandlerInterface.php (hunk 13 above)
Comment (cableman) [reply src/Service/ImageHandlerInterface.php:49]: Renamed

File: src/Service/Indexing/IndexingOrganization.php (hunk 22 above)
Comment (cableman) [reply src/Service/Indexing/IndexingOrganization.php:97]: I like the idea, but most of this is searchable at ES in php... so this in php matches the docs at ES.

**Pull request merged**